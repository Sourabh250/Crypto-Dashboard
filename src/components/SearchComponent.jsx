import React, { useState, useEffect } from "react";
import { FaSearch } from "react-icons/fa";
import { debounce } from "lodash";
import { useDispatch } from "react-redux";

// Custom hook for fetching search suggestions
function useSearchData(query) {
    const [suggestions, setSuggestions] = useState([]);

    useEffect(() => {
        const fetchData = debounce(() => {
            if (query.length > 1) {
                fetch(`https://api.coingecko.com/api/v3/search?query=${query}`)
                    .then(response => response.json())
                    .then(data => {
                        setSuggestions(data.coins);
                    })
                    .catch(error => console.log("Error fetching data:", error));
            } else {
                setSuggestions([]);
            }
        }, 500);

        fetchData();

        return () => {
            fetchData.cancel(); // Cancel debounce on unmount
        };

    }, [query]);
    return suggestions;
}

function SearchComponent() {
    const dispatcher = useDispatch();
    const [val, setVal] = useState("");
    const [showData, setShowData] = useState(false);
    const suggestions = useSearchData(val); // Get search suggestions based on input

    // Handle input change
    const handleChange = (e) => {
        setVal(e.target.value);
        setShowData(true);
    }

    // Handle suggestion click
    const handleClick = (id) => {
        console.log(id);
        dispatcher({type: "MODIFY_CRYPTO_CURRENCY", payload: id.charAt(0).toUpperCase() + id.slice(1)});
        setShowData(false);
    }

    return (
        <div className="flex-grow shadow-lg relative">

            {/* Search input field */}
            <input onChange={handleChange} className="w-full px-16 py-4 bg-gray-100 rounded-lg focus:outline-none focus:ring-2 focus:ring-red-400" placeholder="Search by coin" value={val} />

            {/* Search icon */}
            <FaSearch className="absolute left-6 top-1/2 transform -translate-y-1/2 text-2xl opacity-40" />

            {/* Show suggestions */}
            {showData && suggestions.length > 0 && (
                <ul className="absolute bg-gray-300 w-full mt-2 z-30 rounded shadow-md border  border-gray-300 max-h-40 overflow-y-auto custom-scrollbar">
                    {suggestions.map((suggestion,index) => (
                        <li key={index} className="p-2 hover:bg-red-200 cursor-pointer" onClick={() => handleClick(suggestion.id)}>{suggestion.name}</li>
                    ))}
                </ul>
            )}
        </div>
    )
}

export default SearchComponent;