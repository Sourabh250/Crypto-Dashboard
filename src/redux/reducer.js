function reducer(state = {currency : "USD", selectedCurrencies: ["Ethereum"]},action) {
    switch(action.type) {
        // Update the selected currency in the state
        case "SET_CURRENCY":
            return {
                ...state,
                currency: action.payload
            };
        // Add a new cryptocurrency to the list of selected cryptocurrencies
        case "ADD_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: [...state.selectedCurrencies, action.payload],
            };
        // Remove a cryptocurrency from the list of selected cryptocurrencies
        case "REMOVE_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: state.selectedCurrencies.filter(data => data !== action.payload),
            };
        // Modify the list of selected cryptocurrencies to only contain the specified cryptocurrency
        case "MODIFY_CRYPTO_CURRENCY" :
            return {
                ...state,
                selectedCurrencies: [action.payload],
            }
        // Return the current state if the action type is not recognized
        default:
            return state;
    }
}

export default reducer;