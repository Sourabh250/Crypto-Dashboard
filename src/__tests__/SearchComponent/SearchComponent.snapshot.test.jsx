import React from "react";
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from "../../redux/reducer";
import SearchComponent from '../../components/SearchComponent';

const store = createStore(reducer);

test('SearchComponent snapshot', () => {
    const tree = renderer.create(
        <Provider store={store}>
            <SearchComponent />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
})