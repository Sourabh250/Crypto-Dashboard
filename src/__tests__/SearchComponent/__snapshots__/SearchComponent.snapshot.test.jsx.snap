// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent snapshot 1`] = `
<div
  className="flex-grow shadow-lg relative"
>
  <input
    className="w-full px-16 py-4 bg-gray-100 rounded-lg focus:outline-none focus:ring-2 focus:ring-red-400"
    onChange={[Function]}
    placeholder="Search by coin"
    value=""
  />
  <svg
    className="absolute left-6 top-1/2 transform -translate-y-1/2 text-2xl opacity-40"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
    />
  </svg>
</div>
`;
